api/api.go

```diff
@@ -75,226 +77,216 @@ func (api *ImportAPI) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		auditActionFailure(ctx, uploadFileAction, actionSuccessful, auditError, logData)
 		return
 	}
-	log.Info("added uploaded file to job", log.Data{jobIDKey: jobID, "file": uploadedFile})
+
+	audit.LogInfo(ctx, "added uploaded file to job", logData)
 }
 
-func (api *ImportAPI) updateJob(w http.ResponseWriter, r *http.Request) {
+func (api *ImportAPI) updateJobHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
 	vars := mux.Vars(r)
 	jobID := vars["id"]
 	logData := log.Data{jobIDKey: jobID}
 	auditParams := common.Params{jobIDKey: jobID}
 
-	if err := api.auditor.Record(r.Context(), updateJobAction, actionAttempted, auditParams); err != nil {
-		handleAuditError(w, err, logData)
+	if auditError := api.auditor.Record(ctx, updateJobAction, actionAttempted, auditParams); auditError != nil {
+		handleAuditError(ctx, w, updateJobAction, actionAttempted, auditError, logData)
 		return
 	}
 
-	job, err := models.CreateJob(r.Body)
-	defer r.Body.Close()
-	if err != nil {
-		log.Error(err, logData)
-		http.Error(w, "bad client request received", http.StatusBadRequest)
-		return
-	}
+	options := make(map[string]bool)
+	if err := func(options map[string]bool) (err error) {
+		job, err := models.CreateJob(r.Body)
+		defer r.Body.Close()
+		if err != nil {
+			log.ErrorCtx(ctx, errors.WithMessage(err, "updateJob endpoint: failed to update job resource"), logData)
+			options[badRequest] = true
+			return
+		}
 
-	logData["job"] = job
-	if err := api.jobService.UpdateJob(r.Context(), jobID, job); err != nil {
-		log.ErrorC("error updating job", err, logData)
-		if err := api.auditor.Record(r.Context(), updateJobAction, actionUnsuccessful, auditParams); err != nil {
-			handleAuditError(w, err, logData)
+		logData["job"] = job
+		if err = api.jobService.UpdateJob(ctx, jobID, job); err != nil {
+			log.ErrorCtx(ctx, errors.WithMessage(err, "updateJob endpoint: failed to store updated job resource"), logData)
 			return
 		}
-		setErrorCode(w, err)
+
+		return nil
+	}(options); err != nil {
+		if auditError := api.auditor.Record(ctx, updateJobAction, actionUnsuccessful, auditParams); auditError != nil {
+			handleAuditError(ctx, w, updateJobAction, actionUnsuccessful, auditError, logData)
+			return
+		}
+		setErrorCode(w, err, options)
 		return
 	}
 
-	if err := api.auditor.Record(r.Context(), updateJobAction, actionSuccessful, auditParams); err != nil {
-		handleAuditError(w, err, logData)
+	if auditError := api.auditor.Record(ctx, updateJobAction, actionSuccessful, auditParams); auditError != nil {
+		auditActionFailure(ctx, updateJobAction, actionSuccessful, auditError, logData)
 		return
 	}
-	log.Info("job update completed successfully", logData)
+
+	audit.LogInfo(ctx, "job update completed successfully", logData)
 }
 
-func setErrorCode(w http.ResponseWriter, err error) {
+func setErrorCode(w http.ResponseWriter, err error, opts map[string]bool) {
```

Comment:
why is opts a map[string]bool? Wouldn't it be much simpler if it was a bool?

Reply:
future proofing

Reply:
Reluctant 👍